docs: #862 HA guidelines (#891)
redpanda-data/documentation: docs/platform/deployment/high-availability.mdx

- `acks=1`: The producer waits for an acknowledgment from the leader, but it doesn’t wait for the leader to get acknowledgments from replicas. This setting doesn’t prioritize throughput, latency, or durability. Instead, `acks=1` attempts to provide a balance between all of them.
- `acks=all`: The producer receives an acknowledgment after the leader and the majority of (and therefore, implicitly, all) replicas acknowledge the message. This increases durability at the expense of lower throughput and increased latency.

### Partition rebalancing

By default, Redpanda rebalances partition distribution when nodes are added or decommissioned. Continuous Data Balancing additionally rebalances partitions when nodes become unavailable or when disk space usage exceeds a threshold.

See also: [Cluster Balancing](../../cluster-administration/cluster-balancing/)

### Tiered Storage and disaster recovery

Ensure that you have enough space, that your storage is as fast as possible, and that you use an [XFS filesystem](../../introduction/autotune/#disk). 

> In a disaster, your HA cluster may still be available, but you need to quickly return to the state where redundancy can again save your cluster. HA deployments should use Tiered Storage for infinite data retention and topic recovery. If a broker only has local storage and is running in a containerized environment like Kubernetes, all state is lost for this broker when the Pod is removed. 

Reviewer: ```suggestion
In a disaster, your secondary cluster may still be available, but you need to quickly restore the original level of redundancy by bringing up a new primary cluster. In a containerized environment such as Kubernetes, all state is lost from pods that use only local storage. HA deployments with Tiered Storage address both these problems, since it offers infinite data retention and topic recovery.
```

## Failure scenarios

Failures can happen, even on virtualized infrastructure. The following table shows common types of failures and the likely impact on system availability:

|  Failure     | Impact | Mitigation |
| ----------- | ----------- | ----------- |
| Node failure | Loss of function for an individual node or for VMs hosted on that node | Clustered deployment |
| Rack or switch failure | Loss of nodes/VMs hosted within that rack, or loss of connectivity to them | Clustered deployment spread across multiple racks or network failure domains |
| DC/DC-room failure | Loss of nodes/VMs hosted within that DC/DC-room, or loss of connectivity to them | Stretch cluster or replicated deployment |
| Region failure | Loss of nodes/VMs hosted within that region, or loss of connectivity to them | Geo stretched-cluster (latency dependent) or replicated deployment |
| Global, systemic outage (DNS failures, routing failures) | Complete outage for all systems and services impacting customers and staff | Offline backups, replicas in 3rd-party domains |

> | Human action (inadvertent or malicious) | Corrupt availability of data and any synchronous replicas | Offline backups |

Reviewer: How about emphasizing "data loss" instead of "human action", since data loss is the actual problem and human action is the cause of the problem? Also, "corrupt availability of data" means the availability is corrupt, but what we really mean is that the data is corrupt or unavailable (as a result of accidental deletion or a malicious act).
```suggestion
| Data loss or corruption (accidental or malicious) | Corrupt or unavailable data that also affects synchronous replicas | Offline backups |
```
Question for @tmgstevens and @vuldin:
For the mitigation, could we mention Tiered Storage? I was talking to Josh earlier about possibly promoting the use of Tiered Storage as an important part of a DR plan (making frequent "back-ups" to Tiered Storage to achieve an RPO objective).

Reviewer 2: I think it would be a good idea to mention tiered storage here, and then also follow up in the future with a DR-focused blog detailing how tiered storage fits into threes plans.

Author: added

## HA deployments

> You have different options for business-critical deployments of Redpanda, and there are tradeoffs with the options. For some scenarios, an active/passive deployment is appropriate. Other scenarios trade latency for geo-resilience. 

Reviewer: Suggestion:
Add a bullet list of the topics that follow, to summarize the different types of deployment options.

### Multi-region deployment

A multi-region deployment looks similar to a multi-AZ deployment. It has at least three regions to counter the loss of a single region. With this deployment, there are two ways to retain produce and fetch latency:
- The first option is to manually configure leadership of each partition to ensure that leaders are congregated in the primary region (closest to the producers and consumers). 
- The second option is to configure producers to have `acks=1`; however, this introduces the possibility of losing messages if the primary region becomes lost or unavailable.

### Multi-cluster deployment
  
In a multi-cluster deployment, each cluster can be configured with one of other HA deployment scenarios, but standby or Remote Read Replica clusters are available in one or more remote locations. One of the following asynchronous replication options is required:

> - MirrorMaker (simple producer/consumer-based replication) or MirrorMaker2 (preserves offsets between replicas)

Author: Do we still support MirrorMaker v1?

Reviewer 2: I don't know that any testing has been done with MM1, but the familiar "it should work" applies here.

Reviewer 3: Should we link to the following page to provide more info on using MM2 with Redpanda?
https://docs.redpanda.com/docs/platform/data-management/data-migration/

Reviewer: Since we don't want to be "on the hook" for supporting the old MirrorMaker, can we just omit the reference to MM and focus on MM2?

The leader considers a replica to be inconsistent if it fails to send an acknowledgment. In some cases, it may be acceptable to have inconsistent replicas. For example, a topic could have enough partitions spread across brokers that one unresponsive broker is acceptable. Configure this by setting `min-insync.replicas` to the number of replicas that must be in-sync.

While Apache Kafka® uses in-sync replicas, Redpanda uses a quorum-based majority. This is because Redpanda uses the Raft replication protocol. While Kafka performance is negatively impacted when any replica is out-of-sync, a Redpanda cluster tolerates replica failures without any performance degradation.  

Monitor the health of your cluster with the [`rpk cluster health`](../../reference/rpk/#rpk-cluster-health) command. This tells you how many nodes are down (if any), as well as if you have any leaderless partitions. 

### Rack awareness

Rack awareness is one of the most important features for HA. It lets Redpanda spread partition replicas across available brokers in a way that avoids keeping those replicas in the same physical location. 

:::tip

> Rack location is up to the user, so make sure that separate rack IDs actually lead to a physical separation of brokers. 

Reviewer: ```suggestion
Assign separate rack IDs that actually correspond to a physical separation of brokers. 
```

:::

For more information, see [Rack Awareness](../../data-management/rack-awareness/).

> ### Partition leadership

Author: The new Redpanda architecture page has new doc on partition leadership elections. I'll link to that when available.

### Producer acknowledgment

Producer acknowledgment defines how producer clients and broker leaders communicate their status while transferring data. The following `acks` values determine producer and broker behavior when writing data to the event bus: 
- `acks=0`: The producer doesn’t wait for acknowledgments from the leader broker and doesn’t retry sending messages. This increases throughput and lowers latency of the system at the expense of durability.
- `acks=1`: The producer waits for an acknowledgment from the leader broker, but it doesn’t wait for the leader to get acknowledgments from replicas. This setting doesn’t prioritize throughput, latency, or durability. Instead, `acks=1` attempts to provide a balance between all of them.
- `acks=all`: The producer receives an acknowledgment after the leader and all replicas acknowledge the message. This increases durability at the expense of lower throughput and increased latency.

> ### Partition rebalancing

Author: SMEs: Please confirm that the content here is the latest info on cluster balancing!

Reviewer 2: These details look correct, and I don't know of any recent changes that impact this.

Author: @vuldin It's the first paragraph here that I'm questioning: "Rebalancing can also be triggered when a new node is detected by setting enable_auto_rebalance_on_node_add to true."
I believe now `partition_autobalancing_mode` is set to `node_add` by default. What do you think about replacing that first paragraph with some of the text from the [Cluster Balancing page](https://deploy-preview-891--redpanda-documentation.netlify.app/docs/platform/cluster-administration/cluster-balancing/), like "By default, Redpanda rebalances partition distribution when nodes are added or decommissioned. Continuous Data Balancing additionally rebalances partitions when nodes become unavailable or when disk space usage exceeds a threshold."?

Reviewer 2: I think that would be a good change to the first paragraph, thanks @micheleRP .

- vectorized_raft_configuration_change_in_progress: Indicates if current Raft group configuration is in joint state (i.e. configuration is being changed)

See also: [Cluster Balancing](../../cluster-administration/cluster-balancing/)

### Storage

High availability deployments should use remote storage. If a broker only has local storage and is running in a containerized environment like Kubernetes, all state is lost for this broker when the pod is removed.

Also ensure that you have enough space, that your storage is as fast as possible, and that you use an [XFS filesystem](../../introduction/autotune/#disk).

### Disaster recovery

Have a plan for how to recover from disasters. Your HA cluster may still be available, but you need to quickly return to the state where redundancy can again save your cluster. Redpanda Tiered Storage backs up partitions to any S3-compatible location. Tiered Storage includes:
- Topic recovery
- Remote Read Replicas
- Infinite data retention

See also: [Tiered Storage](../../data-management/tiered-storage/)

> 

Reviewer 4: Is there a "Next steps" aspect here? This doc describes potential failure scenarios. Seems like we should have a next steps heading that points them to docs to troubleshoot any of these scenarios. Even if this represents a content gap and the doc for addressing any of these failures is forthcoming, consider placing that heading here and populating it with the appropriate links when that other aspect of disaster recovery is documented.

Author: @Feediver1 This is a conceptual topic, so I don't think there's any next step. I'll update this page when the the next HA blog is ready (see https://github.com/redpanda-data/documentation/issues/838), which may necessitate a what's next note.

| Data loss or corruption (accidental or malicious) | Corrupt or unavailable data that also affects synchronous replicas | Offline backups (Tiered Storage) |

## HA deployments

You have different options for mission-critical deployments of Redpanda, and there are tradeoffs with the options. For some scenarios, an active/passive deployment is appropriate. Other scenarios trade latency for geo-resilience. 

- [Clustered deployment](../../deployment/high-availability/#clustered-deployment)
- [Multi-AZ deployment](../../deployment/high-availability/#multi-availability-zone-deployment)
- [Multi-region deployment](../../deployment/high-availability/#multi-region-deployment)
- [Multi-cluster deployment](../../deployment/high-availability/#multi-cluster-deployment)

### Clustered deployment

Redpanda is designed to be deployed in a cluster and can scale to many nodes. (Although clusters with a single node work fine, they aren’t resilient to failure.) [Rack awareness](../../deployment/high-availability/#rack-awareness) ensures that Redpanda can tolerate the loss of a rack or failure domain.

### Multi-availability zone deployment
  

> An availability zone (AZ) is a discrete data center with discrete failure domains (power, cooling, fire, network) served by high-bandwidth links with low latency (and typically within a close distance of one another). AZs still have common-cause failure domains, such as force majeure events. A cluster can be deployed across multiple AZs by configuring each AZ as a rack using rack awareness. 

Reviewer: An AZ consists of one or more AZs, according to [AWS docs](https://aws.amazon.com/about-aws/global-infrastructure/regions_az/), so I'm suggesting that minor change. Also, here's a rewrite that emphasizes that multi-AZs are the "solution", or safeguard. See what you think!

How about:
```suggestion
An availability zone (AZ) consists of one or more data centers served by high-bandwidth links with low latency (and typically within a close distance of one another). All AZs have discrete failure domains (power, cooling, fire, and network), but they also have common-cause failure domains such as catastrophic events that affect their geographical location. To safeguard against such possibilities, a cluster can be deployed across multiple AZs by configuring each AZ as a rack using rack awareness. 
```

> An availability zone (AZ) is a discrete data center with discrete failure domains (power, cooling, fire, network) served by high-bandwidth links with low latency (and typically within a close distance of one another). AZs still have common-cause failure domains, such as force majeure events. A cluster can be deployed across multiple AZs by configuring each AZ as a rack using rack awareness. 

Author: > Can we add a link for the definition of force majeure (https://en.wikipedia.org/wiki/Force_majeure)

We reworded paragraph to avoid the term altogether: https://deploy-preview-891--redpanda-documentation.netlify.app/docs/platform/deployment/high-availability/#multi-availability-zone-deployment

- [Replica synchronization](../../deployment/high-availability/#replica-synchronization)
- [Rack awareness](../../deployment/high-availability/#rack-awareness)
- [Partition leadership](../../deployment/high-availability/#partition-leadership)
- [Producer acknowledgment](../../deployment/high-availability/#producer-acknowledgment)
- [Partition rebalancing](../../deployment/high-availability/#partition-rebalancing)
- [Tiered Storage and disaster recovery](../../deployment/high-availability/#tiered-storage-and-disaster-recovery)

### Replica synchronization

A cluster’s availability is directly tied to replica synchronization. Brokers can be either leaders or replicas (followers) for a partition. A cluster’s brokers are available for clients (consumers or producers) only if the brokers are consistent with the leader. 

1. The leader writes data to the log and determines which replica brokers are in-sync. 
2. The replicas receive messages written to the partition of the leader broker and send acknowledgments to the leader after successfully replicating the message to their internal partition. 

> 3. The leader then sends an acknowledgment to the producer of the message, as determined by that producer’s `acks` value.

Reviewer 5: Add in: Redpanda will consider the group to be consistent once a majority has formed consensus (i.e. there is a simple majority of participants having acknowledged the write).

> The leader considers a replica to be inconsistent if it fails to send an acknowledgment. In some cases, it may be acceptable to have inconsistent replicas. For example, a topic could have enough partitions spread across brokers that one unresponsive broker is acceptable. Configure this by setting `min-insync.replicas` to the number of replicas that must be in-sync.

Reviewer 5: Remove this.

If the leader becomes unresponsive before the replicas have fully synchronized, you could have an unclean leader election. This keeps the cluster available, but it comes at the expense of durability (potential loss of data). An alternative is to not allow unclean leader election, but that could lead to cluster failure.  

### Producer acknowledgment

Producer acknowledgment defines how producer clients and broker leaders communicate their status while transferring data. The following `acks` values determine producer and broker behavior when writing data to the event bus.
- `acks=0`: The producer doesn’t wait for acknowledgments from the leader broker and doesn’t retry sending messages. This increases throughput and lowers latency of the system at the expense of durability.
- `acks=1`: The producer waits for an acknowledgment from the leader broker, but it doesn’t wait for the leader to get acknowledgments from replicas. This setting doesn’t prioritize throughput, latency, or durability. Instead, `acks=1` attempts to provide a balance between all of them.

> - `acks=all`: The producer receives an acknowledgment after the leader and all replicas acknowledge the message. This increases durability at the expense of lower throughput and increased latency.

Reviewer 5: Change "all" to "the majority of (and therefore, implicitly, all"

Author: fixed
